### PR TITLE
Model sampling/text generation

### PR DESCRIPTION
**Pull request by joshlk**
Text generation is working with no output (unconditional) and with an input file. By "working" I mean it is loading a model, creating output and saving it to file. The output is currently utter gibberish as I'm testing it on basically a randomly initialised model.

Steps to test:

**Generate a model:**
1. first change the `save-interval` to 10 or something small in the model config.
2. Run a model (note this now saves the model config in the checkpoint directory so you can easily determine what the model data corresponds to):

`./deepy.py pretrain_gpt2.py -d configs small.yml eleutherai_cluster.yml` and wait for a checkpoint to save.

**To test `unconditional` text generation:**
1. Make sure `"text-gen-type": "unconditional"` is set in the `text_generation.yml` config
2. Run an text generation job loading the generated model:

`./deepy.py text_gen_gpt2.py  -d configs small.yml eleutherai_cluster.yml text_generation.yml`

3. Output will be found in `/mnt/ssd-cluster/output/text_generation.txt`

**To test `input-file` text generation:**
1. Make sure `"text-gen-type": "input-file"` is set in the `text_generation.yml` config
2. Create an new line delineated input file at `/mnt/ssd-cluster/output/sample_input.txt`
2. Run a text generation job:

`./deepy.py text_gen_gpt2.py  -d configs small.yml eleutherai_cluster.yml text_generation.yml`

3. Output will be found in `/mnt/ssd-cluster/output/sample_output.txt`

Todo:
- [X] Model loading
- [X] Unconditional output
- [X] Output from file
- [X] Save model config to checkpoint folder
- [x] Create example script in examples folder

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -7,7 +7,7 @@ RUN apt-get update -y && \
     apt-get install -y \
         git python3 python3-dev libpython3-dev python3-pip sudo pdsh \
         htop llvm-9-dev tmux zstd software-properties-common build-essential autotools-dev \
-        nfs-common pdsh cmake g++ gcc curl wget tmux less unzip htop iftop iotop ca-certificates ssh \
+        nfs-common pdsh cmake g++ gcc curl wget vim less unzip htop iftop iotop ca-certificates ssh \
```
Comment (joshlk) [inline Dockerfile:10]: Note: I haven't removed `tmux`, it was just specified twice and added `vim` (which used to be added by default in an older image)

File: deepy.py [hunk 2]
```diff
@@ -51,14 +52,19 @@ def get_wandb_api_key():
...
     os.makedirs(conf['log-dir'], exist_ok=True)
     file_prefix = os.path.join(conf['log-dir'], '0-deepy')
     Tee(file_prefix+'_stdout.txt', err=False)
     Tee(file_prefix + '_stderr.txt', err=True)
 
+if 'save' in conf:
+    os.makedirs(conf['save'], exist_ok=True)
+    config_file = os.path.join(conf['save'], 'config.yml')
+    with open(config_file, 'w') as f:
+        json.dump(conf, f, indent=4)
```
Comment (joshlk) [inline deepy.py:67]: Here I save the config file into the checkpoint directory

File: megatron/text_generation_utils.py [hunk 3]
```diff
@@ -95,71 +94,76 @@ def generate_samples_input_from_file(model):
...
-            for _, decode_tokens in enumerate(token_stream):
-                decode_tokens, _ = decode_tokens
-                decode_tokens = decode_tokens[0].cpu().numpy().tolist()
+        for tokens, length in zip(token_batch, length_batch):
+            tokens = tokens[1:length - 1]
+            try:
+                text = tokenizer.detokenize(tokens)
+            except KeyError:
+                print_rank_0("WARNING: generated token which doesn't exist. Skipping")
+                continue
```
[the remark this answers is not in the document]
Comment (StellaAthena) [reply megatron/text_generation_utils.py:152]: When we run the model, it says that it needs to create dummy tokens. Could that be connected?

[the remark this answers is not in the document]
Comment (StellaAthena) [reply megatron/text_generation_utils.py:152]: When we run the model, it says that it needs to create dummy tokens. Could that be connected?

**Review by StellaAthena — DISMISSED** [review]
This code requires significant changes to ensure it is up-to-date with `main` before it can be merged.

**Comment by sdtblck** [comment]
I've brought the branch up to date with main and am now beginning to test

**Comment by sdtblck** [comment]
We default pipe-parallel-size to 0 when sampling, but this is only working for me when i *train* at pipe-parallel-size = 0.

So we need to find a way to either:

- Convert PP model's weights to MP
- Sample with a PP model

**Comment by sdtblck** [comment]
Ok, this is now working with pipeline parallel size > 1.

The merge conflicts will be quite intense (since rotary pos emb means the number of args passed between pipe stages can differ). I'll sort these soon and we can merge this.

**Comment by sdtblck** [comment]
Ok so sampling is now working with all pipeline parallel sizes. It's the most awful, hacky code ever, you need to update your deeperspeed branch to the latest commit to get it to work. This should be ready to be merged now imo.

**Review by ShivanshuPurohit — APPROVED** [review]

**Review by ShivanshuPurohit — DISMISSED** [review]

**Review by sdtblck — APPROVED** [review]

**Review by ShivanshuPurohit — DISMISSED** [review]

**Review by sdtblck — APPROVED** [review]

**Pull request merged**